Command: remove slashes from commands

abeatrix opened the pull request:
CLOSE https://github.com/sourcegraph/cody/issues/1730

Remove slashes from command names and implemented design from https://www.figma.com/file/jEsnzgsf0hNuJbqB9pKaGm/VS-Code---Cody---Commands-up-front-and-ChatGPT%2B%2B?type=design&node-id=1550-62492&mode=design&t=qdjbLWsl9qHq578p-4

Main Commands Menu

![image](https://github.com/sourcegraph/cody/assets/68532117/ec5c6a1b-84a8-4403-82e8-c6b51bd80967)

Custom Commands Menu

![image](https://github.com/sourcegraph/cody/assets/68532117/95552be6-11a4-47ab-8565-cd0db8267a74)

![image](https://github.com/sourcegraph/cody/assets/68532117/34e4dea5-99e0-40b8-b7a0-8676ab199373)

![image](https://github.com/sourcegraph/cody/assets/68532117/d56dc295-6ade-44e0-9b09-bec59c470eed)



## Test plan

<!-- Required. See https://sourcegraph.com/docs/dev/background-information/testing_principles. -->

Updated all current tests.


https://github.com/sourcegraph/cody/assets/68532117/581ef1e9-c0bc-461d-83ae-ffc03098d2fc


### Before 

![image](https://github.com/sourcegraph/cody/assets/68532117/5428e444-922c-4ea2-b75d-28741668b884)


![image](https://github.com/sourcegraph/cody/assets/68532117/f2ea3d94-18a3-4a5a-bb93-a72e326da7f9)

valerybugakov commented on vscode/src/commands/CommandsController.ts:
```diff
@@ -34,7 +35,7 @@ class CommandsController implements vscode.Disposable {
         // The unique key for the command. e.g. /test
         const commandKey = commandSplit?.shift() || text
-        const command = this.provider?.get(commandKey)
+        const command = this.provider?.get(fromSlashCommand(commandKey))
```
Is this required for backward compat?

valerybugakov commented on vscode/src/commands/menus/index.ts:
```diff
@@ -25,7 +26,7 @@ export async function showCommandMenu(
             items.push(CommandMenuSeperator.commands)
             for (const [_name, _command] of vscodeDefaultCommands) {
-                const label = _command.slashCommand
+                const label = fromSlashCommand(_command.slashCommand)
```
Does it make sense to deprecate `CodyCommand.slashCommand` and introduce `CodyCommand.command` or smth like this to reflect the updated behavior better?

abeatrix replied:
good point, i've deprecated slashCommand and replace it with key instead

toolmantim commented:
Design wise we want to have it match the commands treeview.

Could we update both the treeview and the command menu to match this?
https://www.figma.com/file/jEsnzgsf0hNuJbqB9pKaGm/VS-Code---Cody---Commands-up-front-and-ChatGPT%2B%2B?type=design&node-id=1550-62492&mode=design&t=qdjbLWsl9qHq578p-4

abeatrix commented:
@toolmantim updated and ready for review

toolmantim reviewed (commented):
This is great!

A few small changes if we could:

* "New Chat" from the quickpick should act the same as "New Chat" in the review (open a new chat tab)
* Could we add the keyboard shortcuts for "New Chat" and "Edit Code" to the quickpick item descriptions?
* If there are no custom commands, can we make "Custom Commands" quickpick item & treeview item go directly to the "Configure Custom Commands…" and skip the empty-ish middle quickpick?

One bug:

* In the quickpick "Custom Commands" > "Configure Custom Commands…" both the "New Custom Command..." and "Open Custom Commands Documentation" items don't seem to work

abeatrix commented:
@toolmantim updated based on your suggestions and fixed the bug:
- [x] "New Chat" from the quickpick should act the same as "New Chat" in the review (open a new chat tab)
- [x] Could we add the keyboard shortcuts for "New Chat" and "Edit Code" to the quickpick item descriptions?
  - [x] Also added a keyboard shortcuts for "Custom Commands" as shown in your figma design
- [x] If there are no custom commands, can we make "Custom Commands" quickpick item & treeview item go directly to the "Configure Custom Commands…" and skip the empty-ish middle quickpick?
- [x] Fixed bug: In the quickpick "Custom Commands" > "Configure Custom Commands…" both the "New Custom Command..." and "Open Custom Commands Documentation" items don't seem to work

toolmantim reviewed (approved):
Nice! Two non-blocking things…

If you start typing "New" you end up seeing duplicates. I'd suggest we always exclude the "New Chat (shortcut)" and "Edit Code (shortcut)" items when you start filtering/typing… because they're always going to be there as "Start a new chat" and "Start a new edit" anyhow.

![CleanShot 2024-02-13 at 16 18 16@2x](https://github.com/sourcegraph/cody/assets/153/b6e5bbb3-76b9-4b1a-b986-f56c99024540)

Also when you filter in custom commands we probably don't want to show these last 3 items, just the filtered custom commands (or nothing if no matches):

![CleanShot 2024-02-13 at 16 18 40@2x](https://github.com/sourcegraph/cody/assets/153/d2b065f8-6993-413e-84fc-7e532b9d285f)

toolmantim commented:
Does this binding actually exist? I can't get it to trigger:

![CleanShot 2024-02-13 at 16 24 19@2x](https://github.com/sourcegraph/cody/assets/153/bbe7732b-63f2-4ced-9aac-fef138592d5f)

If it doesn't exist, we can just remove that shortcut label for now.

abeatrix commented:
> Does this binding actually exist? I can't get it to trigger:
> 
> ![CleanShot 2024-02-13 at 16 24 19@2x](https://private-user-images.githubusercontent.com/153/304302850-bbe7732b-63f2-4ced-9aac-fef138592d5f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc4MDMwNDUsIm5iZiI6MTcwNzgwMjc0NSwicGF0aCI6Ii8xNTMvMzA0MzAyODUwLWJiZTc3MzJiLTYzZjItNGNlZC05YWFjLWZlZjEzODU5MmQ1Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIxM1QwNTM5MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MmRmY2U1NWIxODFmMzFkODVkMDY3N2Y4NDk2Y2I5ZmJiY2RhZWI1MjM5MzU2MTFlZWYyNjE0MDM1NzBkNDhiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.JT7bdkOeFMMPVi4qTQFaAx0HQU9G-oj07z3j5KR2X5g)
> 
> If it doesn't exist, we can just remove that shortcut label for now.

it does exist! I just fixed this and the bugs you found. Will update the e2e tests tomorrow to cover these cases before I finalize the PR tomorrow, thanks for taking the time to review and test this PR out!

valerybugakov commented on vscode/src/commands/index.ts:
```diff
@@ -0,0 +1,82 @@
+import os from 'os'
+import type { CodySidebarTreeItem } from '../services/treeViewItems'
+
+const isMac = os.platform() === 'darwin'
```
Nit: we have multiple helpers like this? Is it worth reusing the same implementation everywhere? 
<img width="394" alt="Screenshot 2024-02-13 at 15 23 57" src="https://github.com/sourcegraph/cody/assets/3846380/1817880c-d5e0-4ad4-8b58-0c95990828c5">

abeatrix replied:
Good catch! I've move the one for node to lib/shared!

valerybugakov commented on vscode/src/commands/menus/types.ts:
```diff
@@ -20,8 +20,15 @@ export interface CommandMenuItem extends QuickPickItem {
     command?: string
-    // cody command, e.g. '/ask'
+    /**
+     * cody command, e.g. '/ask'
+     * @deprecated Use 'commandKey' instead.
+     */
     slashCommand?: string
```
Nice! 🙌

valerybugakov reviewed (approved):
Looks great!
